### PR TITLE
Extend the documentation of PartialOrd

### LINKED ISSUE #50230: Document "latent requirements" of PartialOrd
The `PartialOrd` docs do not document what is de facto required of trait implementors in order for data structures to work correctly. We should update them!

Per @RalfJung's [comment below](https://github.com/rust-lang/rust/issues/50230#issuecomment-846104238):

> FWIW our conclusion was that we basically already require `<` to be irreflexive, since that is the only way the following equivalences can be upheld:
> * `x < y` iff `x.partial_cmp(y) == Some(Less)`
> * `x == y` iff `x.partial_cmp(y) == Some(Equal)`
> * `x > y` iff `x.partial_cmp(y) == Some(Greater)`
>
> The docs are not explicitly stating that these equivalences must hold, but it seems very much against the spirit of this trait to allow violations of these equivalences. The fact that `PartialOrd: PartialEq` is very useful here as it makes the 2nd axiom well-defined.

<details>
<summary>This post used to say something different. Here is the original post</summary>

Why doesn't `PartialOrd` require irreflexivity `!(a < a)` over `<` ? AFAICT `PartialOrd` does not actually express any kind of ordering relation for `<`. For: 

* a strict partial order on `<`: we need `PartialOrd` + `Irreflexivity`
* non-strict partial order on `<=`: we need `PartialOrd` + `Eq` since `==` must be reflexive.
* total order: we need`PartialOrd + Eq + Total`

Is this an oversight? If not, we should document that `PartialOrd` requires:

* irreflexivity: `!(a < a)`

as well. That would make `PartialOrd`'s `<` denote a strict partial order on `<`. 

AFAICT this is actually what was intended, since all the discussions I can find online about `PartialOrd` talk about floats, and floats are strictly partially ordered under `<`, but not non-strictly partially ordered under `<=`, which would fit our current model that floats don't implement `Eq`.

---

Also, the docs of `PartialOrd` do not say what `le` (`<=`) and `ge` (`>=`) actually mean. AFAICT they mean nothing for `PartialOrd` in isolation. From reading the source the default implementation is `a < b || a == b`, which would denote a non-strict total ordering over `<=` when `Eq` is implemented. But since `Eq` is not required for `PartialOrd`, I'd guess that those overriding these implementations are allowed to do something else as long as they don't implement `Eq`. It is unclear to me what that "something else" might mean, but the docs should probably say something about this.

---

Also, I haven't been able to find any document explaining why things are the way they are, but the fact that `PartialOrd` requires `PartialEq` makes no sense to me, and it doesn't make much sense either that `PartialOrd` exposes `le` (`<=`) and `ge` (`>=`) methods, since as mentioned above, without `Eq`, `le` and `ge` do not express any meaningful ordering relation. It seems that these traits, like all other "operators" traits, mix operator overloading with mathematical semantics: in some cases they are used to express mathematical concepts, like a strict partial ordering relation under `<`, and in some other cases they are used "to provide <= for floats". If anybody has any links to any discussions / documentation about why things are the way they are I'd like to read them, and maybe we should add rationale sections to the docs.

</details>

### PR DESCRIPTION
**Pull request by gnzlbg**
Closes #50230 .

cc @ExpHP @RalfJung @LukasKalbertodt @fbstj

### CODE REVIEW

**Comment by gnzlbg** [comment]
I tried to merge all the comments of https://github.com/LukasKalbertodt/partial-cmp-docs/pull/1/files , but there is still one unresolved issue open by @RalfJung :

> And why state irreflexivity and others nit also for >? Also, antisymmetry as given is a property of both < and > in combination, which I don't think is the usual formulation. (Or rather, the usual definition of for > to be the dual of < to asymmetry and antisymmetry as given here are actually the same statement.)

Also, i'm not a native speaker, and this is a tricky topic, so it would be cool if someone from the docs team would take a look at this and further process it for easy consumption by users (cc @steveklabnik )

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -8,17 +8,16 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-//! Functionality for ordering and comparison.
+//! Functionality for ordering and relational operations.
```
Comment (gnzlbg) [inline src/libcore/cmp.rs]: Should this be just relational operations ? This is what comparisons are.

File: src/libcore/cmp.rs [hunk 2]
```diff
@@ -32,6 +31,66 @@
...
+//! This module also provides the following two traits:
+//!
+//! * [`Eq`]: which requires `PartialEq`, states that all values of a type can be
+//!   compared to all others,
+//! * [`Ord`]: which requires `PartialOrd` and `Eq`, states that the ordering
+//!   relations expressed by the comparison operators form all total ordering relations 
+//!   (more on this below).
+//!
+//! The distinction between these traits is important: it tell us what it _means_ for a 
+//! sequence of values to be ordered.
```
Comment (gnzlbg) [inline src/libcore/cmp.rs]: This is the pragmatic way I choose to put it, might not be the best one. 

@LukasKalbertodt I basically decided to put the "intro docs" to `PartialOrd` at the top level of the module. Otherwise the docs of `PatialOrd` and other traits end up restating the same things over and over again and is hard to keep them in sync.

File: src/libcore/cmp.rs [hunk 3]
```diff
@@ -32,6 +31,66 @@
...
+//!
+//! Two commonly encountered examples are integers and floating-point numbers. For example, every [`i32`] 
+//! value is comparable to every other. If you can tell that two `i32` values are distinc, then these two
+//! values never compare equal, and if you sorte a sequence of `i32` values like `[1, 1, 0, -5]` using `<=`,
```
Comment (pitdicker) [inline src/libcore/cmp.rs]: Just a small nit while reading the new docs: sort

File: src/libcore/cmp.rs [hunk 4]
```diff
@@ -32,6 +31,66 @@
...
+//! doing so will always produce the same unique result: `[-5, 0, 1, 1]`. We say that `<=` imposes a _total_
+//! order, and denote this by implementing `Ord` for `i32`.
+//!
+//! However, none of this holds for IEEE-754 floating-point numbers like `f32`. There is a special `f32` value 
+//! called "Not a Number" (`NaN`) for which any relational operation alwys return false, that is:
```
Comment (pitdicker) [inline src/libcore/cmp.rs]: always

**Comment by pietroalbini** [comment]
Ping from triage @steveklabnik! This PR needs your review.

File: src/libcore/cmp.rs [hunk 5]
```diff
@@ -32,6 +31,66 @@
...
+//! Two commonly encountered examples are integers and floating-point numbers. For example, every [`i32`] 
+//! value is comparable to every other. If you can tell that two `i32` values are distinc, then these two
+//! values never compare equal, and if you sorte a sequence of `i32` values like `[1, 1, 0, -5]` using `<=`,
+//! doing so will always produce the same unique result: `[-5, 0, 1, 1]`. We say that `<=` imposes a _total_
+//! order, and denote this by implementing `Ord` for `i32`.
+//!
+//! However, none of this holds for IEEE-754 floating-point numbers like `f32`. There is a special `f32` value 
+//! called "Not a Number" (`NaN`) for which any relational operation alwys return false, that is: 
+//! `a == NaN` is always `false`, even when `a` is `NaN`. For this reason `f32` only implements `PartialOrd`, and it 
+//! does not implement neither `Eq` nor `Ord`. What does this mean in practice? It means two things:
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: The sentence is wrong somehow. Maybe put it like this:

> For this reason, `f32` implements `PartialOrd` and `PartialEq`, but neither `Ord` nor `Eq`.

File: src/libcore/cmp.rs [hunk 6]
```diff
@@ -32,6 +31,66 @@
...
+//!
+//! * there are sequences of `f32`s that we can order using `<=` and will always produce the same unique 
+//!   result: sorting `[3.0, 5.0, 1.0]` into `[1.0, 3.0, 5.0]`.
+//!
+//! * there are also sequences of `f32`s that when ordered using `<=` will produce a different result depending
+//!  on which algorithm we use to sort them: we can sort `[-0.0, NaN, +0.0] into all its permutations 
+//! (`[-0.0, +0.0, NaN]`, `[+0.0, -0.0, NaN]`, `[NaN, +0.0, -0.0]`, etc. are all "sorted" under `<=`).
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: I don't really like this part (the two bullet points) :/

I think I would replace everything from "What does this mean in practice?" to "are all "sorted" under `<=`)." with something like:

> While this is no problem for sequences like `[3.0, 5.0, 1.0]`, there is no clear order for sequences containing `NaN`: what is the correct way to sort `[-1.0, NaN, 4.0]`? The order `[NaN, -1.0, 4.0]` would be exactly as correct as `[-1.0, 4.0, NaN]`. So there is no unique ordering for floats.

Also: "are all "sorted" under `<=`" -> is that even correct? I'm not quite sure what you mean, but `NaN <= +0.0` is `false`.

File: src/libcore/cmp.rs [hunk 7]
```diff
@@ -32,6 +31,66 @@
...
+//! Another example is the "subset relation" on sets in which, for sets `a` and `b`, `a < b` is true if and only 
+//! if `a` is a proper subset of `b` (e.g. `{4, 9} < {1, 4, 5, 9}` is true). The problem is that when neither 
+//! `a` is a subset of `b` nor `b` is a subset of `a` nor `a = b` (e.g. `{1, 3}` and `{1, 4}`), we cannot sensible 
+//! compare `a` and `b`, so that `a < b`, `a == b`, and `a > b` all return false.
+//!
+//! The distinction between a _total_ order and a _partial_ order is important, and impacts what it means for
+//! the values of a type to be "sorted". This impacts how it makes sense to use these values. For example,
+//! if we sort a vector of `f32`s containing `NaN`s, does it make sense to use binary search to try to find a value
+//! in that vector? Will it always work? Does it make sense to put `f32`s containing `NaN`s into a set? a hash table?
+//! etc. Algorithms and data-structures that require a total order or total equality to work correctly can constrain
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: I would replace "into a set? a hash table? etc." with "into a set or hash table?"

File: src/libcore/cmp.rs [hunk 8]
```diff
@@ -32,6 +31,66 @@
...
+//! the types they accept with the [`Eq`] and [`Ord`] traits, while those for which a partial order or partial equality
+//! is enough can use the `PartialEq` and `PartialOrd` traits.
+//!
+//! This leads us to the final question: which traits should a type implement ? There is no easy answer: a type should only
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: Nit: space before `?`. Also: the lines are pretty long here.

File: src/libcore/cmp.rs [hunk 9]
```diff
@@ -512,16 +571,56 @@ impl PartialOrd for Ordering {
     }
 }
 
-/// Trait for values that can be compared for a sort-order.
+/// Trait for values that can be partially ordered.
+///
+/// This traits defines the operators `<`, `>`, `<=`, and `>=` which compare two values and return `bool`, 
+/// where `<` and `>` form [_strict_ partial ordering relations][wiki_porder].
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: I would split this sentence. "[...] return `bool`. The operators `<` and `>` form [...]"

File: src/libcore/cmp.rs [hunk 10]
```diff
@@ -512,16 +571,56 @@ impl PartialOrd for Ordering {
...
+///
+/// The `PartialOrd` trait  has fewer requirements of the ordering relation than Ord. In particular, `partial_cmp`
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: Nits: two spaces after `trait` and `Ord` not in monofont

File: src/libcore/cmp.rs [hunk 11]
```diff
@@ -512,16 +571,56 @@ impl PartialOrd for Ordering {
...
+/// * _duality_: if `a < b` then `b > a`
+///
+/// The `lt` (`<`), `le` (`<=`), `gt` (`>`), and `ge` (`>=`) methods are implemented in terms of `partial_cmp` 
+/// according to these rules. The default implementations can be overridden for performance reasons, but manual 
+/// implementations must satisfy the rules above.
+///
+/// From these rules it follows that `PartialOrd` must be implemented symmetrically and transitively.
+/// That is, for two distinct types `T` and `U`, `T: PartialOrd<U>` requires the following trait 
+/// implementations to exist and satisfy these rules: `T: PartialOrd<T>`, `U: PartialOrd<T>`, 
+/// and `U: PartialOrd<U>`. If for a third distinct type `V` `T: PartialOrd<V>` is provided, then
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: Mhhh I don't quite get why `T: PartialOrd<T>` and `U: PartialOrd<U>` are required from the rules above. I am probably just missing something obvious...

File: src/libcore/cmp.rs [hunk 12]
```diff
@@ -512,16 +571,56 @@ impl PartialOrd for Ordering {
...
+/// in terms of the same `partial_cmp`:
+///
+/// * irreflexivity of `<`: `!(a < a)`
+/// * transitivity of `>`: if `a > b` and `b > c` then `a > c`
+/// * asymmetry of `<`: if `a < b` then `!(b < a)`
+///
+/// Stronger ordering relations can be expressed by using the `Eq` and `Ord` traits, where the `PartialOrd` 
+/// methods provide:
+///
+/// * a [_non-strict_ partial ordering relation](https://en.wikipedia.org/wiki/Partially_ordered_set#Strict_and_non-strict_partial_orders)
```
Comment (LukasKalbertodt) [inline src/libcore/cmp.rs]: Did you mean to use `[wiki_porder]` here? Since you added the link at the bottom...

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
I finally managed to review this. Sorry for the delay!

And thanks a lot for pushing this topic.

I left a few comments. Many are just typos or minor things. However, as you can see below, I'm not quite happy with changes in the module docs -- but maybe it's just me. 

Also note that all "I would..." comments should be taken with a grain of salt since I'm not a native speaker.

Oh and maybe you could summarize the prior discussion and what this PR mainly wants to achieve so that @steveklabnik can start reviewing this more quickly.

File: src/libcore/cmp.rs (hunk 11 above)
Comment (gnzlbg) [reply src/libcore/cmp.rs]: This is required by transitivity (if `a < b` and `b < c` then `a < c`).

Suppose that `a: T`, `b: U`, and `c: T`., and that `a < b` and `b < c` are both true. For `a < c` to also be true it would need to be well-typed, and that requires `T: PartialOrd<T>`.

Comment (LukasKalbertodt) [reply src/libcore/cmp.rs]: Ah right, that makes sense. Thanks :)

File: src/libcore/cmp.rs (hunk 6 above)
Comment (gnzlbg) [reply src/libcore/cmp.rs]: Yes I wasn't fully happy with this either. For me, all are sorted under `<=`  means that `!(a > b)` is true for all adjacent pairs in the sequence. When either `a` or `b` are `NaN` that's automatically satisfied. Sadly for floats that's not exactly the same as `a <= b` being `true` for all adjacent pairs...

**Comment by gnzlbg** [comment]
cc @RalfJung could you comment on these changes? I will then go and update the PR with both your feedback, if you agree with @LukasKalbertodt feedback a 👍 on their comments would be great!

File: src/libcore/cmp.rs [hunk 13]
```diff
@@ -8,17 +8,16 @@
...
+//! This module defines four traits related to ordering and comparisons:
 //!
-//! [`PartialOrd`]: trait.PartialOrd.html
-//! [`PartialEq`]: trait.PartialEq.html
+//! * [`PartialEq`]: which overloads the `==` and `!=` operators,
+//! * [`PartialOrd`] (requires `PartialEq`): which overloads the `<`, `<=`, `>`, 
+//!   and `>=` operators,
 //!
-//! # Examples
+//! state that values of a type can be compared for equality, or ordered, but that
```
Comment (RalfJung) [inline src/libcore/cmp.rs]: I find this hard to parse, where does the sentence even begin and what is the subject matching this verb?

File: src/libcore/cmp.rs [hunk 14]
```diff
@@ -32,6 +31,66 @@
 //! assert_eq!(x == y, false);
 //! assert_eq!(x.eq(&y), false);
 //! ```
+//!
+//! This module also provides the following two traits:
+//!
+//! * [`Eq`]: which requires `PartialEq`, states that all values of a type can be
+//!   compared to all others,
+//! * [`Ord`]: which requires `PartialOrd` and `Eq`, states that the ordering
+//!   relations expressed by the comparison operators form all total ordering relations
```
Comment (RalfJung) [inline src/libcore/cmp.rs]: What does it mean to form "all total ordering relations"?

Or do you mean, "all the comparison operators form total ordering relations"?

File: src/libcore/cmp.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/cmp.rs]: The second part of this sentence does not talk about a distinction, so I am confused what this is even telling us.

File: src/libcore/cmp.rs [hunk 15]
```diff
@@ -32,6 +31,66 @@
...
+//! values never compare equal, and if you sorte a sequence of `i32` values like `[1, 1, 0, -5]` using `<=`,
+//! doing so will always produce the same unique result: `[-5, 0, 1, 1]`. We say that `<=` imposes a _total_
+//! order, and denote this by implementing `Ord` for `i32`.
+//!
+//! However, none of this holds for IEEE-754 floating-point numbers like `f32`. There is a special `f32` value 
+//! called "Not a Number" (`NaN`) for which any relational operation alwys return false, that is: 
+//! `a == NaN` is always `false`, even when `a` is `NaN`. For this reason `f32` only implements `PartialOrd`, and it 
+//! does not implement neither `Eq` nor `Ord`. What does this mean in practice? It means two things:
+//!
+//! * there are sequences of `f32`s that we can order using `<=` and will always produce the same unique
```
Comment (RalfJung) [inline src/libcore/cmp.rs]: "There"

File: src/libcore/cmp.rs [hunk 16]
```diff
@@ -32,6 +31,66 @@
...
+//!
+//! However, none of this holds for IEEE-754 floating-point numbers like `f32`. There is a special `f32` value 
+//! called "Not a Number" (`NaN`) for which any relational operation alwys return false, that is: 
+//! `a == NaN` is always `false`, even when `a` is `NaN`. For this reason `f32` only implements `PartialOrd`, and it 
+//! does not implement neither `Eq` nor `Ord`. What does this mean in practice? It means two things:
+//!
+//! * there are sequences of `f32`s that we can order using `<=` and will always produce the same unique 
+//!   result: sorting `[3.0, 5.0, 1.0]` into `[1.0, 3.0, 5.0]`.
+//!
+//! * there are also sequences of `f32`s that when ordered using `<=` will produce a different result depending
```
Comment (RalfJung) [inline src/libcore/cmp.rs]: "There"

File: src/libcore/cmp.rs [hunk 17]
```diff
@@ -512,16 +571,56 @@ impl PartialOrd for Ordering {
...
 ///
-/// - antisymmetry: if `a < b` then `!(a > b)`, as well as `a > b` implying `!(a < b)`; and
-/// - transitivity: `a < b` and `b < c` implies `a < c`. The same must hold for both `==` and `>`.
+/// The comparisons must satisfy, for all `a`, `b`, and `c`:
 ///
-/// Note that these requirements mean that the trait itself must be implemented symmetrically and
-/// transitively: if `T: PartialOrd<U>` and `U: PartialOrd<V>` then `U: PartialOrd<T>` and `T:
-/// PartialOrd<V>`.
+/// * _transitivity_: if `a < b` and `b < c` then `a < c`
+/// * _duality_: if `a < b` then `b > a`
```
Comment (RalfJung) [inline src/libcore/cmp.rs]: Shouldn't it say somewhere that consistency with `==` from `PartialEq` is required?

Basically, this trait contains *strictly more* information than `PartialEq`, and it must be the case that `==` behaves exactly as if implemented via `partial_cmp == Some(Equal)`.

File: src/libcore/cmp.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/cmp.rs]: Actually I am not even sure which distinction you mean: `PartialEq` vs `Eq`, or `Eq` vs `Ord`...

File: src/libcore/cmp.rs (hunk 9 above)
Comment (RalfJung) [reply src/libcore/cmp.rs]: I'd move the remark about strict ordering relations far down to all the corollaries.

**Comment by RalfJung** [comment]
I feel like the module introduction spends lots of text explaining things most users will not find relevant... all this array sorting and so on.

I would start by introducing the 4 traits, and then explain that the `Partial*` variants exist because there are types that to not satisfy all the usual properties we expect from `Eq`: For example, floating points, and finite sets. This shouldn't take nearly as much text as you have here now.

---

For `PartialOrd` (and this is a concern I already raised earlier but it seems it got lost), I'd focus everything around `partial_cmp`. This is the one thing the user is providing, and it contains all the information. So, as the introducing, I'd write

> This traits defines the operators `<`, `>`, `<=`, and `>=` which compare two values and return `bool`. All of these operations are defined in terms of a single function, `partial_cmp(a, b)`, which says how `a` and `b` are related: One can be smaller than the other, they can be equal -- and they can be unrelated. This is also the key difference to `Ord`, which does not allow elements to be unrelated.

> With `partial_cmp` defined, it is easy to define all six comparison operators:
> * `a < b` if and only if `partial_cmp(a, b) == Some(Less)`
> * `a > b` if and only if `partial_cmp(a, b) == Some(Greater)`
> * `a == b` if and only if `partial_cmp(a, b) == Some(Equal)`
> * `a != b` if and only if `!(a == b)`
> * `a <= b` if and only if `a < b || a == b`
> * `a >= b` if and only if `a > b || a == b`

> However, not every choice of `partial_cmp` makes sense. The following must be satisfied for all `a`, `b`, and `c`:

[transitivity and duality go here]

> Moreover, `partial_cmp` must be consistent with `PartialEq`: `==` and `!=` as defined above in terms of `partial_cmp` must produce the same results as the `eq` and `ne` methods from `PartialEq`. Finally, the `lt` (`<`), `le` (`<=`), `gt` (`>`), and `ge` (`>=`) methods of `PartialEq` must be consistent with the definition of the corresponding relations in terms of `partial_cmp` (this is the case for the default implementations, but if you choose to overwrite those for performance reasons, then it is your responsibility to uphold this guarantee).

Then go on about various corollaries (my [open issue](https://github.com/rust-lang/rust/pull/53386#issuecomment-413172653) would be resolved by remarking that transitivity and duality still hold after swapping `<` and `>`), and partial strict ordering relations (Wikipedia link), and whatever else you think is worth mentioning.

**Comment by TimNN** [comment]
Ping from triage, @gnzlbg: Skimming the comments above, I believe this PR requires an update from you.

**Comment by TimNN** [comment]
Ping from triage @gnzlbg: it looks like this PR needs to be updated.

**Comment by TimNN** [comment]
Ping from triage @gnzlbg: I'm closing this due to inactivity per the [triage procedure](https://forge.rust-lang.org/triage-procedure.html). Thank you for your contribution and please feel free to (re-)open this or another PR in the future.
